### PR TITLE
Initial S3 destination

### PR DESCRIPTION
**Pull request by jsantos17**
[ch8288]

### CODE REVIEW

File: core/src/main/scala/quasar/destination/s3/S3DestinationModule.scala [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        case (e: S3Exception) if Eq[Int].eqv(403, e.statusCode) =>
+          DestinationError.accessDenied((
+            destinationType,
+            originalConfig,
+            "Access denied")).asLeft
+      }
+
+  private def mkClient[F[_]: Sync](cfg: S3Config): Resource[F, S3AsyncClient] = {
+    val client =
+      Sync[F].delay(
```
Comment (jsantos17) [inline core/src/main/scala/quasar/destination/s3/S3DestinationModule.scala]: I tried to configure a custom `ExecutorService` here, but it doesn't seem to work. I'm still seeing the default threads all around in addition to threads from my executor. [Thread dump](https://gist.github.com/jsantos17/e4049c163f6e2a5d2da70ee274dfa8b1). Note now we have both `s3-async-dest-1` and `aws-java-sdk-NettyEventLoop-0-7` 😕

You can find the implementation with a custom executor service [here](https://github.com/jsantos17/quasar-destination-s3/blob/ch8288-s3-destination-custom-executor/core/src/main/scala/quasar/destination/s3/S3DestinationModule.scala#L108).

File: build.sbt [hunk 2]
```diff
@@ -21,5 +30,24 @@ lazy val core = project
   .settings(name := "quasar-destination-s3")
   .settings(
     performMavenCentralSync := false,
+    quasarPluginName := "s3-dest",
+    quasarPluginQuasarVersion := QuasarVersion,
+    quasarPluginDestinationFqcn := Some("quasar.destination.s3.S3DestinationModule$"),
+    quasarPluginDependencies ++= Seq(
+      "io.argonaut"  %% "argonaut" % ArgonautVersion,
+      "co.fs2" %% "fs2-core" % Fs2Version,
+      "io.monix" %% "monix-catnap" % MonixVersion,
```
Comment (jsantos17) [inline build.sbt]: There's a bit of dependency problem here. `monix-catnap` is only available for cats-effect > 2.0, but quasar is still on cats-effect 1.4.0. It's causing unsafe evictions. I'm thinking about dropping `monix-catnap` and instead one-off it here. So far I haven't seen any problems related to the cats-effect eviction.

Comment (djspiewak) [reply build.sbt]: Cats Effect 2 is binary compatible with Cats Effect 1. Cats as well. I'm considering just bumping us across the board.

File: core/src/main/scala/quasar/destination/s3/impl/DefaultUpload.scala [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  CompleteMultipartUploadRequest,
+  CompleteMultipartUploadResponse,
+  CompletedMultipartUpload,
+  CompletedPart,
+  CreateMultipartUploadRequest,
+  CreateMultipartUploadResponse,
+  UploadPartRequest
+}
+
+final case class DefaultUpload[F[_]: Concurrent](client: S3AsyncClient, partSize: Int) extends Upload[F] {
```
Comment (djspiewak) [inline core/src/main/scala/quasar/destination/s3/impl/DefaultUpload.scala]: The fact that there's only a `Concurrent` bound here indicates that the thread shifting is not being done correctly. In other words, the thread pools inside of the AWS client are leaking into sdbe. In order to fix this, a `ContextShift` is required (note the difference between `IO.fromFuture` in Cats Effect 1.0 and our `fromFutureShift` function).

File: core/src/main/scala/quasar/destination/s3/impl/DefaultUpload.scala [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    client: S3AsyncClient,
+    bytes: Stream[F, Byte],
+    uploadId: String,
+    minChunkSize: Int,
+    bucket: Bucket,
+    key: ObjectKey): F[List[CompletedPart]] =
+    (bytes.chunkMin(minChunkSize).zipWithIndex evalMap {
+      case (byteChunk, n) => {
+        // parts numbers must start at 1
+        val partNumber = Int.box(n.toInt + 1)
```
Comment (djspiewak) [inline core/src/main/scala/quasar/destination/s3/impl/DefaultUpload.scala]: Why the explicit boxing?

File: core/src/main/scala/quasar/destination/s3/impl/DefaultUpload.scala [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    }).compile.toList
+
+  private def completeUpload(
+    client: S3AsyncClient,
+    uploadId: String,
+    bucket: Bucket,
+    key: ObjectKey,
+    parts: List[CompletedPart]): F[CompleteMultipartUploadResponse] = {
+    val multipartUpload =
+      CompletedMultipartUpload.builder.parts(parts :_*).build
```
Comment (djspiewak) [inline core/src/main/scala/quasar/destination/s3/impl/DefaultUpload.scala]: I generally prefer `parts: _*` here, to be uniform with conventional ascription styling.

File: .gitignore [hunk 6]
```diff
@@ -15,3 +15,5 @@ target/
 /pubring.pgp.enc
 /scripts/readVersion
 /secring.pgp.enc
+
+testCredentials.json
```
Comment (djspiewak) [inline .gitignore]: You probably want `/testCredentials.json` here

File: core/src/main/scala/quasar/destination/s3/impl/DefaultUpload.scala (hunk 5 above)
Comment (jsantos17) [reply core/src/main/scala/quasar/destination/s3/impl/DefaultUpload.scala]: Ohh I didn't see splats as type ascription. But makes sense when you put it that way.

File: core/src/main/scala/quasar/destination/s3/impl/DefaultUpload.scala (hunk 4 above)
Comment (jsantos17) [reply core/src/main/scala/quasar/destination/s3/impl/DefaultUpload.scala]: Not sure. Implicit conversions are triggered when I import `scala.Predef._`, but don't work when I use `slamdata.Predef._`.

Comment (jsantos17) [reply core/src/main/scala/quasar/destination/s3/impl/DefaultUpload.scala]: Hmm. Seems like `int2Integer` is available in `scala.Predef` but not in `slamdata.Predef`

File: core/src/main/scala/quasar/destination/s3/S3DestinationModule.scala [hunk 7]
```diff
@@ -79,9 +79,9 @@ object S3DestinationModule extends DestinationModule {
     } yield (S3Destination(cfg.bucket, upload): Destination[F])).value
   }
 
-  private def isLive[F[_]: Async](client: S3AsyncClient, originalConfig: Json, bucket: Bucket)
+  private def isLive[F[_]: Concurrent](client: S3AsyncClient, originalConfig: Json, bucket: Bucket)
       : F[Either[InitializationError[Json], Unit]] =
-    Async[F].delay(client.headBucket(HeadBucketRequest.builder.bucket(bucket.value).build))
+    Concurrent[F].delay(client.headBucket(HeadBucketRequest.builder.bucket(bucket.value).build))
```
Comment (djspiewak) [inline core/src/main/scala/quasar/destination/s3/S3DestinationModule.scala]: This doesn't actually change anything

File: core/src/main/scala/quasar/destination/s3/S3DestinationModule.scala [hunk 8]
```diff
@@ -97,9 +97,9 @@ object S3DestinationModule extends DestinationModule {
             "Access denied")).asLeft
       }
 
-  private def mkClient[F[_]: Sync](cfg: S3Config): Resource[F, S3AsyncClient] = {
+  private def mkClient[F[_]: Concurrent](cfg: S3Config): Resource[F, S3AsyncClient] = {
     val client =
-      Sync[F].delay(
+      Concurrent[F].delay(
```
Comment (djspiewak) [inline core/src/main/scala/quasar/destination/s3/S3DestinationModule.scala]: Neither does this

File: build.sbt [hunk 9]
```diff
@@ -21,5 +30,24 @@ lazy val core = project
   .settings(name := "quasar-destination-s3")
   .settings(
     performMavenCentralSync := false,
+    quasarPluginName := "s3-dest",
```
Comment (jsantos17) [inline build.sbt]: Worth noting this is called `s3-dest` rather than `s3`. In order to avoid conflicts with the datasource.

**Pull request merged**